src/index.js

```diff
@@ -0,0 +1,224 @@
+import { promisify } from 'util';
+
+const glob = promisify(globCb);
+
+zkSnark.bigInt.prototype.toJSON = function() {
+  return this.toString();
+};
+
+const Extension = {
+  Json: '.json',
+  Circom: '.circom',
+  VkProof: '.vk_proof',
+  VkVerifier: '.vk_verifier',
+  Solidity: '.sol'
+};
+
+class Snarks {
+  constructor(embark) {
+    this.embark = embark;
+    this.circuitsConfig = embark.pluginConfig;
+    this.compiledCircuitsPath = embark.config.dappPath('.embark', 'snarks');
+    // v5 detection won't work as desired with "dev embark" (built in monorepo)
+    // until after first prerelease of embark v5; can assign `true` to test
+    // this.isEmbark5 = true;
+    this.isEmbark5 = semver(this.embark.version).major >= 5;
+    this.registerEvents();
+  }
+
+  static circomBinary = findUp('node_modules/.bin/circom', { cwd: __dirname });
+  static snarkjsBinary = findUp('node_modules/.bin/snarkjs', {
+    cwd: __dirname
+  });
+
+  registerEvents() {
+    if (this.isEmbark5) {
+      return this.embark.registerActionForEvent(
+        'compiler:contracts:compile:before',
+        (contractFiles, callback) => {
+          this.compileAndGenerateContracts(contractFiles, callback);
+        }
+      );
+    }
+    this.embark.registerActionForEvent('build:beforeAll', callback =>
+      this.compileAndGenerateContracts([], callback)
+    );
+  }
+
+  verifierFilepath(basename) {
+    return path.join(
+      this.compiledCircuitsPath,
+      `${basename}${Extension.VkVerifier}`
+    );
+  }
+
+  verifierContractPath(basename) {
+    return path.join(
+      this.compiledCircuitsPath,
+      `${basename}${Extension.Solidity}`
+    );
+  }
+
+  proofFilepath(basename) {
+    return path.join(
+      this.compiledCircuitsPath,
+      `${basename}${Extension.VkProof}`
+    );
+  }
+
+  async compileCircuits() {
+    await fs.ensureDir(this.compiledCircuitsPath);
+
+    const patterns = this.circuitsConfig.circuits;
+    if (!patterns) {
+      return;
+    }
+
+    this.embark.logger.info('Compiling circuits...');
+
+    await Promise.all(
+      patterns.map(async pattern => {
+        const filepaths = await glob(pattern);
+        await Promise.all(
+          filepaths
+            .filter(filename => path.extname(filename) === Extension.Circom)
+            .map(filepath => this.compileCircuit(filepath))
+        );
+      })
+    );
+  }
+
+  compileCircuit(filepath) {
+    return new Promise((resolve, reject) => {
+      const output = path.join(
+        this.compiledCircuitsPath,
+        `${path.basename(filepath, Extension.Circom)}${Extension.Json}`
+      );
+      exec(`${Snarks.circomBinary} ${filepath} -o ${output}`, error => {
+        if (error) {
+          return reject(error);
+        }
+
+        resolve();
+      });
+    });
+  }
+
+  async generateProofs() {
+    this.embark.logger.info('Generating proofs...');
+
+    await Promise.all(
+      (await fs.readdir(this.compiledCircuitsPath))
```

Comment:
Wrapping in parens here is a cool trick to avoid creating an intermediary variable. Didn't know this is a thing!

Reply:
Yes, being able to use `await` with whatever expressions you want is really wonderful and powerful. What's also cool is the same goes for `yield` and `yield*`.